File: 2020/08/halting.hs
{-# LANGUAGE MultiWayIf #-}
{-# LANGUAGE LambdaCase #-}

import Text.Read hiding (step, get)
import Data.Array
import Data.Monoid
import Control.Monad.State

data Instruction = Acc Int | Jmp Int | Nop Int
    deriving (Show)

instance Read Instruction where
    readPrec = do
        Ident op <- lexP
        Symbol sign <- lexP
        n <- readPrec
        let arg = case sign of
                    "+" -> n
                    "-" -> -n
        return $ case op of
                   "acc" -> Acc arg
                   "jmp" -> Jmp arg
                   "nop" -> Nop arg

flipInstruction :: Instruction -> Instruction
flipInstruction (Acc n) = Acc n
flipInstruction (Nop n) = Jmp n
flipInstruction (Jmp n) = Nop n

data Program = Program { instructions :: Array Int Instruction
                       , executed :: Array Int Bool
                       , accumulator :: Int
                       , position :: Int
                       } deriving (Show)

data Status = Running | Loop | Terminated
    deriving (Show, Eq)

main = do insts <- (map read . lines) <$> getContents
          let len = length insts - 1
              pgm = Program (listArray (0, len) insts) (listArray (0, len) $ repeat False) 0 0
          putStrLn . show . accumulator . execState run $ pgm


exec :: Instruction -> State Program ()
exec (Acc n) = modify $ \p -> p { accumulator = accumulator p + n
                                , position = position p + 1
                                }
exec (Nop _) = modify $ \p -> p { position = position p + 1 }
exec (Jmp n) = modify $ \p -> p { position = position p + n }

step :: State Program Status
step = get >>= \p ->
    if | position p > (snd . bounds . instructions $ p) -> return Terminated
       | executed p ! position p -> return Loop
       | otherwise -> do
           put p { executed = executed p // [(position p, True)] }
           exec $ instructions p ! position p
           return Running

run :: State Program Status
run = step >>= \case Running -> run
                     other -> return other

File: 2020/08/halting.rs
use std::{
    io::{self, BufRead},
    str::FromStr,
    error::Error,
    iter::repeat,
    convert::TryInto,
    fmt,
};

#[derive(Debug)]
enum Instruction {
    Acc(i64),
    Jmp(i64),
    Nop(i64),
}

use Instruction::*;

#[derive(Debug)]
struct ParseError;

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Parse error!")
    }
}

impl Error for ParseError {}

impl FromStr for Instruction {
    type Err = Box<dyn Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut words = s.split(' ');
        let op = match words.next() {
            Some("acc") => |n| Acc(n),
            Some("jmp") => |n| Jmp(n),
            Some("nop") => |n| Nop(n),
            Some(_) => return Err(Box::new(ParseError)),
            None => return Err(Box::new(ParseError)),
        };

        return match words.next() {
            Some(x) => match x.parse::<i64>() {
                Ok(n) => Ok(op(n)),
                Err(err) => Err(Box::new(err)),
            },

            None => Err(Box::new(ParseError)),
        };
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
enum Status {
    Running,
    Loop,
    Terminated,
}

use Status::*;

#[derive(Debug, Clone, Copy)]
struct State {
    position: usize,
    accumulator: i64,
    status: Status,
}

#[derive(Debug)]
struct Program {
    instructions: Vec<(Instruction, bool)>,
    state: State,
}

impl Program {
    fn new<I>(ins: I) -> Self
    where
        I: Iterator<Item = Instruction>,
    {
        return Program {
            instructions: ins.zip(repeat(false)).collect(),
            state: State {
                position: 0,
                accumulator: 0,
                status: Running,
            },
        }
    }
}

impl Iterator for Program {
    type Item = State;

    fn next(&mut self) -> Option<Self::Item> {
        if self.state.status != Running {
            return None;
        }

        if self.state.position >= self.instructions.len() {
            self.state.status = Terminated;
            return Some(self.state);
        }

        match self.instructions[self.state.position] {
            (_, true) => {
                self.state.status = Loop;
            }
            (Acc(n), _) => {
                self.state.accumulator += n;
                self.instructions[self.state.position].1 = true;
                self.state.position += 1;
            }
            (Jmp(n), _) => {
                self.instructions[self.state.position].1 = true;
                self.state.position = (self.state.position as i64 + n).try_into().unwrap();
            }
            (Nop(_), _) => {
                self.instructions[self.state.position].1 = true;
                self.state.position += 1;
            }
        }

        return Some(self.state);
    }
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();

    let prog = Program::new(stdin.lock().lines().map(|x| x.unwrap().parse().unwrap()));
    if let Some(state) = prog.last() {
        println!("{}", state.accumulator);
    } else {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "Couldn't run instructions"));
    }

    return Ok(());
}
